Dockerfile

```diff
@@ -0,0 +1,35 @@
+# Copyright (C) Daniel Stenberg, <daniel@haxx.se>, et al.
+#
+# SPDX-License-Identifier: curl
+
+# Self-contained build environment to match the release environment.
+#
+# Build with
+#
+#   docker build -t curl/curl .
+#
+# Then run commands from within the build environment, for example
+#
+#   docker run --rm -v $(pwd):/usr/src -w /usr/src curl/curl autoreconf -fi
+#   docker run --rm -v $(pwd):/usr/src -w /usr/src curl/curl ./configure --without-ssl --without-libpsl
+#   docker run --rm -v $(pwd):/usr/src -w /usr/src curl/curl make
+#   docker run --rm -v $(pwd):/usr/src -w /usr/src curl/curl ./maketgz 8.7.1
+
+FROM debian:bookworm-slim@sha256:993f5593466f84c9200e3e877ab5902dfc0e4a792f291c25c365dbe89833411f
+# autoconf 2.71
+# automake 1.16.5
+# libtoolize 2.4.7
+# make 4.3
+# perl 5.36.0
+# git 2.39.2
+
+RUN apt-get update -qq && apt-get install -qq -y --no-install-recommends \
```

Comment:
If we want to deterministically install the same package versions I think we need to drop the `apt-get update` here? Otherwise you'll get whatever version was latest in bookworm at the time you run the docker build command. Which is not reproducibile.. Instead update the `FROM` clause whenever we want to bump the versions for reproducibility.

Reply:
1. This still relies in some non-reproducible packages in debian.
2. You are going to still need tools that don't ship with the minimal debian container
3. Packages installed via apt-get update get purged regularly so you would need to use the debian archive
4. The debian archive is very slow, very unreliable, and very bandwidth constrained. You also need to write a lot of custom tooling to hash-lock exact package versions or apt -will- try to grab whatever the mirror hands it.
5. Signatures on packages pinned today will expire and break apt

I have gone all the way down this road, and realized Debian is just a non-starter for practical/maintainable reproducible builds. I am about to push another PR with a stagex-based Containerfile that is deterministic and full source bootstrapped.

Reply:
Can we be more concrete here? It would be fine if the tool versions differ as long as they still produce the same tarball. Which of the steps taken to produce the tarballs will be give different results when the tool versions differ?

Reply:
A different version of automake autoconf tar, zip, bzip2 etc could all end up causing different results in the future.

I just pushed a draft PR with an example that uses a "from scratch" container and an explicit set of fixed/deterministic/multi-signed dependencies:

https://github.com/curl/curl/pull/13338

Does not have zip support (until next week as we need to do a release cycle to package it) but otherwise should be good to go, and can be easily upgraded to build any deterministic binaries of curl if desired as well.

Reply:
How about something along the lines of this?
```
FROM debian:bookworm-slim@sha256:993f5593466f84c9200e3e877ab5902dfc0e4a792f291c25c365dbe89833411f

RUN apt-get update -qq && apt-get install -qq -y --no-install-recommends \
    build-essential=12.9 \
    make=4.3-4.1 \
    autoconf=2.71-3 \
    automake=1:1.16.5-1.3 \
    ...
```

Reply:
I once asked that same exact question. It led me down a several month rabbit hole I would save you from if I can.

1. sub-dependencies could still shift to latest versions where supported unless you do this for -all- of them, thus you end up needing tools to generate a huge lockfile of every single package like this: https://git.distrust.co/public/airgap/src/branch/main/config/toolchain/packages-x86_64.list
2. As soon as newer versions of dependencies come out, the ones you are pinning now will get pulled from mirrors, so you need to have scripts to automatically pull, pin, then pivot to using the debian archive like this: https://codeberg.org/distrust/toolchain/src/commit/2560e543a1871b86cb1c29b31ce9141eaec94e8c/scripts/packages-update
3. Next you will find the signing keys expire every day, and apt will refuse to install those packages. You can disable signing in apt, but then you -must- implent your own hash locking strategy. You end up having to automatically generate a hashlock file like this: https://git.distrust.co/public/airgap/src/branch/main/config/toolchain/package-hashes-x86_64.txt
4. You will then realize that the debian archive provide files at dialup speeds, if it works at all, so you will need to mirror several GB of old versions of packages yourself with Git LFS or similar.
5. Then you will realize updating all of this is a real pain, and you need to make a whole repo of make helpers and scripts to maintain it all: https://git.distrust.co/public/toolchain
6. Then you realize that all of this -works- but is awful to audit and maintain and no one wants touch projects that use it (I deployed this setup for a few orgs as it was the -only- option for reproducible builds from a mostly reproducible/signed lniux distro.)
7. You then face facts that debian only ships very old versions of packages, and in most cases single maintainers have God keys and complete trust to ship whatever they want, and many critical packages are still not reproducible so trust in the whole tree still comes down to trusting any one out of 100+ people is not compromised or malicious.

Reply:
@lyxell That would work temporarily, but would not remain deterministic as specific Debian package versions get purged and moved to archive. Then you have to either download and maintain the install files so that you still have them available, or switch to using archive links which are painfully slow. You also inherit a ton of dependencies from Debian that you really don't need. The PR @lrvick suggested is easier to maintain, has a smaller dependency footprint, and all of the deps that you do use are built deterministically and multi-reproduced. Just for full transparency I contribute to [StageX](https://codeberg.org/stagex/stagex) which Lance suggested in his PR.

Reply:
I can attest to what Lance is saying. I went through the process of making a build deterministic using Debian and it was a painful process. StageX makes it a lot easier, and the other benefit that hasn't been mentioned is that the whole StageX build-chain is verifiable, deterministic, and multi-reproduced. So what you end up with is that the tool-chain, the packages, and then your software that's built with it are all deterministic, and you get to decide if you want to also multi-reproduce what's being built using this setup.

Reply:
I can imagine that building binaries in a deterministic fashion on Debian will not work in practice. However, I imagine that building tarballs where all of the files of the tarballs will be plain text files should be feasible even if the tool versions used to build those tarballs may vary.

That is, in practice I would expect tools like `perl`, `gzip` and `tar` to produce byte-by-byte identical outputs for the commands listed in `maketgz` the last decade or so on Debian, under all versions. But this is just me guessing, I would be happy to be proven wrong.

Reply:
If you only run `maketgz` as it is implemented right now that would likely work as you described. Side note: `gzip` has likely had optimizations over the years. `automake` changes its mind on stuff from one version to the next so that concern is valid. The natural follow up to this is to have `make` in the curl repo use the container to actually produce deterministic binaries of curl as a baseline and that has to use stagex or something like it to achieve determinism.

Reply:
[ We do deterministic binaries in the sub-project curl-for-win: https://github.com/curl/curl-for-win

It is based on Debian versioned images, with the caveat discussed above, where `apt update`/`install` may break strict reproducibility of the tools used. Code dependencies are all built from source reproducibly. ]

Reply:
The only package versions that matter at all here are the versions of the autotools packages (autoconf, automake, libtool...).

Simply print the debian (apt) versions of this exceedingly small number of packages to a reproducible-manifest.txt inside the tarball before you tar it up. Problem solved.

@lrvick As far as rebuilding all packages from scratch goes, it appears you have reinvented the Gentoo Linux distro except NIH. Which, incidentally, also allows you to record an exact git repository hash of the Gentoo tree and guarantee all packages in your container use that and nothing else.

But all this is the mootest of moot points, because if you want to guarantee that everyone is using the exact same docker container to reproduce the tarballs, then you simply publish the container you used -- as mentioned above. This is *very* reproducible as you're uploading @bagder's actual (virtual) laptop used to make the release, and anyone can *download @bagder's (virtual) laptop and use it to create the tarball* (and check that it's the virtual laptop created in CI).

Reply:
A decision has already been made here so I am merely responding to some potential misconceptions above for the sake of anyone following this thread.

> Simply print the debian (apt) versions of this exceedingly small number of packages to a reproducible-manifest.txt inside the tarball before you tar it up. Problem solved.

That will require the user that wishes to verify to track down all those versions, and all their dependencies, and assemble them together again. Version numbers are not good enough, as those keys expire and are rotated often. You would at a minimum need to publish the entire installed apt tree in the container with /hashes/ of every .deb file so they can be retrieved from the Debian archive and directly installed in a reproduction container. This will be very arduous for a user still without any scripted help, but it would at least be a path. I have written a few tools to help with this mentioned elsewhere. It is a terrible reproduction path, but it does work.

> @lrvick As far as rebuilding all packages from scratch goes, it appears you have reinvented the Gentoo Linux distro except NIH.

Gentoo linux is not reproducible, quorum-signed, full-source-bootstrapped, or OCI native. It is true that we do in fact compile everything from source, but we publish our binaries and expect users to use those, so that makes us closer to traditional linux distros like Debian than Gentoo.

The only other deterministic full-source-bootstrapped Linux distro that exists to my knowledge is GNU Guix.They however do not publish reproducible container images, and do not do quorum signing. They are also glibc based and optimized for end user desktop use, which comes with dramatically higher attack surface and complexity than what is required with a goal of simply securely building software. Still, Guix would be the best distro to compare us to.

If you are not familiar with full-source-bootstrapping, Guix has a great writeup as they were the first distro to do it. Stagex was the second.

https://guix.gnu.org/en/blog/2023/the-full-source-bootstrap-building-from-source-all-the-way-down/
 
> If you want to guarantee that everyone is using the exact same docker container to reproduce the tarballs, then you simply publish the container you used 

Agreed. That is the reason to use containers, but the context of this are trying to minimize the chance of supply chain attacks. We are trying to solve for trust. 

If the binary container image is compromised by the person with the publishing API key, then anything downstream from that container image is also compromised. See the XZ attack and Ken Thompson's Reflections on Trusting Trust.

You can however pull down the "musl" and "autoconf" binary stagex container images and directly use them together. You can also verify multiple people built them for you and got the same exact hash, via their PGP signatures so you don't have to build it if you trust they are not colluding. You also can -choose- to build them yourself and add your own signature if you don't trust those folks.

With Debian OTOH, they build their published docker images in a non-deterministic way and do not sign them, so you have no evidence it is free of tampering. Very much like the XZ situation. Using Docker Debian unfortunately just moves the problem up one layer.

Reply:
All of this is completely irrelevant, since the curl project is trying to solve reproducible curl release tarballs, not reproducible linux distros as general-purpose computing platforms.

> That will require the user that wishes to verify to track down all those versions, and all their dependencies, and assemble them together again. Version numbers are not good enough, as those keys expire and are rotated often. You would at a minimum need to publish the entire installed apt tree in the container with /hashes/ of every .deb file so they can be retrieved from the Debian archive and directly installed in a reproduction container.

No, again, version numbers of the autotools set of packages alone (and not any dependencies) are sufficient since they allow one to determine the correct version of autotools packages to install from ANY source, modulo debian-specific patches publicly recorded in the Debian VCS for that small handful of packages.

No hashes necessary, no dependencies necessary, no access to .debs necessary.

It is slightly more awkward to install a local autotools toolchain with debian patches on your choice of personally trusted computing systems than to install a prebuilt .deb, but not much harder.

Also "as those keys expire and are rotated often" does not sound like a practical objection to me. If you use old stuff signed by crypto keys which are old, then naturally you backtrace the trust source for those keys, and subsequently backdate the verification routine to respect keys that claim to have expired in the past, but from the perspective of what you're reproducing, in the future.

> If you are not familiar with full-source-bootstrapping, Guix has a great writeup as they were the first distro to do it. Stagex was the second.

Sure I'm familiar with it. I don't think that's the concern the curl project has. It will also be more interesting to *me personally* when it no longer relies on untrusted binaries such as idk, the docker daemon, guile-bootstrap, or any other potentially malicious tools used to host the computing platform that runs the full-source bootstrap.

> If the binary container image is compromised by the person with the publishing API key, then anything downstream from that container image is also compromised. See the XZ attack and Ken Thompson's Reflections on Trusting Trust.

I've seen both quite well, thanks. Assuming a non-compromised docker daemon, it really is not hard to rebuild the container image using, say, https://snapshot.debian.org/ and test that your binary container image produces the same tarball. Again, the container doesn't have to be byte-for-byte reproducible, it simply has to be some form of instructions that others can follow to yield the same... curl release tarball.

No need to trust curl devs to publish a non-compromised spinoff container.

> With Debian OTOH, they build their published docker images in a non-deterministic way and do not sign them, so you have no evidence it is free of tampering. Very much like the XZ situation. Using Docker Debian unfortunately just moves the problem up one layer.

Sounds like a market opportunity for someone to build deterministic debian docker images that are signed.

The underlying .deb packages are reproducible: https://tests.reproducible-builds.org/debian/reproducible.html

(Feel free to run your own builders to verify that!)

Reply:
> Also "as those keys expire and are rotated often" does not sound like a practical objection to me. If you use old stuff signed by crypto keys which are old, then naturally you backtrace the trust source for those keys, and subsequently backdate the verification routine to respect keys that claim to have expired in the past, but from the perspective of what you're reproducing, in the future.

If keys are revoked because they used weak, vulnerable, algos or are already compromised, or they were by a maintainer that is no longer trusted (like the xz maintainer) then those keys could be used to sign a malicious alternative to the package you are actually requesting, with the same version number, and apt would not know the difference. This is why hash locking is an absolute necessity when using old packages where the signatures may no longer be trustworthy.

> It really is not hard to rebuild the container image using, say, https://snapshot.debian.org/

This actually is surprisingly hard. Debian shapshots are hosted by a single deployment that is most of the time crippled to dialup speeds in my experience, and the experience of virtually every user that has used projects of mine that relied on it. Also, for the above reasons about not being able to trust expired/revoked keys, you have to pull the .deb files manually without installing, hash verify them, then install from that folder as a local mirror. For these two reasons, in the many reproducible build projects I worked with based on Debian, we all ended up forced to use our own git LFS mirrors of .deb packages which was a major pain.

> It is slightly more awkward to install a local autotools toolchain with debian patches on your choice of personally trusted computing systems than to install a prebuilt .deb, but not much harder.

To be fair to this point, for your very specific use case here, you could probably get away with using generic latest debian and then build your own autotools, and then use your own autotools to build the tar. That is of course if any signed/trusted/reproducible debian base images actually existed :/